File: src/main/java/com/mycompany/sapo_leyendo/service/ShippingService.java
package com.mycompany.sapo_leyendo.service;

import com.mycompany.sapo_leyendo.model.*;
import com.mycompany.sapo_leyendo.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class ShippingService {

    @Autowired
    private TransportLoadRepository transportLoadRepository;

    @Autowired
    private CarrierRepository carrierRepository;

    @Autowired
    private ShipmentRepository shipmentRepository;

    @Autowired
    private ManifestRepository manifestRepository;
    
    @Autowired
    private OutboundOrderRepository outboundOrderRepository;

    @Transactional
    public TransportLoad createLoad(Integer carrierId, String trailerNumber, String driverName, String driverPhone) {
        Carrier carrier = carrierRepository.findById(carrierId)
                .orElseThrow(() -> new RuntimeException("Carrier not found"));

        TransportLoad load = new TransportLoad();
        load.setLoadNumber("LOAD-" + System.currentTimeMillis());
        load.setCarrier(carrier);
        load.setTrailerNumber(trailerNumber);
        load.setDriverName(driverName);
        load.setDriverPhone(driverPhone);
        load.setStatus(LoadStatus.PLANNING);
        load.setCreatedAt(LocalDateTime.now());
        
        return transportLoadRepository.save(load);
    }

    @Transactional
    public void assignShipmentToLoad(Integer loadId, Integer shipmentId) {
        TransportLoad load = transportLoadRepository.findById(loadId)
                .orElseThrow(() -> new RuntimeException("Load not found"));

        if (load.getStatus() == LoadStatus.IN_TRANSIT || load.getStatus() == LoadStatus.DELIVERED) {
            throw new RuntimeException("Cannot assign shipment to a dispatched or delivered load");
        }

        Shipment shipment = shipmentRepository.findById(shipmentId)
                .orElseThrow(() -> new RuntimeException("Shipment not found"));

        shipment.setTransportLoad(load);
        shipmentRepository.save(shipment);

        if (load.getStatus() == LoadStatus.PLANNING) {
            load.setStatus(LoadStatus.LOADING);
            transportLoadRepository.save(load);
        }
    }

    @Transactional
    public Manifest dispatchLoad(Integer loadId) {
        TransportLoad load = transportLoadRepository.findById(loadId)
                .orElseThrow(() -> new RuntimeException("Load not found"));

        if (load.getStatus() == LoadStatus.IN_TRANSIT) {
            throw new RuntimeException("Load is already dispatched");
        }

        List<Shipment> shipments = load.getShipments();
        if (shipments == null || shipments.isEmpty()) {
            throw new RuntimeException("Cannot dispatch an empty load");
        }

        for (Shipment shipment : shipments) {
            shipment.setStatus(ShipmentStatus.SHIPPED);
            shipment.setShippedAt(LocalDateTime.now()); 
            
            // Update Order Status
            OutboundOrder order = shipment.getOutboundOrder();
            if (order != null) {
                order.setStatus("SHIPPED");
                outboundOrderRepository.save(order);
            }
            
            shipmentRepository.save(shipment);
        }

        load.setStatus(LoadStatus.IN_TRANSIT);
        load.setActualDeparture(LocalDateTime.now());
        transportLoadRepository.save(load);

        Manifest manifest = new Manifest();
        manifest.setManifestNumber("MAN-" + System.currentTimeMillis());
        manifest.setTransportLoad(load);
        manifest.setCreatedAt(LocalDateTime.now());

        return manifestRepository.save(manifest);
    }

    // ===== CARRIER OPERATIONS =====
    
    public List<Carrier> getAllCarriers() {
        return carrierRepository.findAll();
    }

    public java.util.Optional<Carrier> getCarrierById(Integer id) {
        return carrierRepository.findById(id);
    }

    // ===== SHIPMENT OPERATIONS =====
    
    public List<Shipment> getAllShipments() {
        return shipmentRepository.findAll();
    }

    public java.util.Optional<Shipment> getShipmentById(Integer id) {
        return shipmentRepository.findById(id);
    }

    @Transactional
    public Shipment createShipment(Integer outboundOrderId, Integer carrierId, String trackingNumber) {
        OutboundOrder order = outboundOrderRepository.findById(outboundOrderId)
                .orElseThrow(() -> new RuntimeException("Order not found: " + outboundOrderId));

        Shipment shipment = new Shipment();
        shipment.setOutboundOrder(order);
        shipment.setCarrierId(carrierId);
        shipment.setTrackingNumber(trackingNumber != null ? trackingNumber : generateTrackingNumber());
        shipment.setStatus(ShipmentStatus.PACKING);
        
        return shipmentRepository.save(shipment);
    }

    @Transactional
    public java.util.Optional<Shipment> updateShipment(Integer id, Shipment updated) {
        return shipmentRepository.findById(id)
                .map(existing -> {
                    if (updated.getCarrierId() != null) existing.setCarrierId(updated.getCarrierId());
                    if (updated.getTrackingNumber() != null) existing.setTrackingNumber(updated.getTrackingNumber());
                    if (updated.getStatus() != null) existing.setStatus(updated.getStatus());
                    if (updated.getTotalWeightKg() != null) existing.setTotalWeightKg(updated.getTotalWeightKg());
                    return shipmentRepository.save(existing);
                });
    }

    public List<TransportLoad> getAllLoads() {
        return transportLoadRepository.findAll();
    }

    private String generateTrackingNumber() {
        return "TRK-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
    }

    /**
     * Rate Shopping Logic (Mock)
     * Finds the cheapest carrier for a given shipment.
     */
    public Carrier findBestCarrier(Double weight, String destinationZip) {
        List<Carrier> carriers = carrierRepository.findAll();
        
        // Mock logic: Randomly select one or pick based on simple rule
        // Real logic would call Carrier APIs
        return carriers.stream()
                .findAny()
                .orElseThrow(() -> new RuntimeException("No carriers available"));
    }
}

File: Visualization/src/services/shippingService.ts
import axios from 'axios';

const API_URL = '/api/shipping';

export interface Carrier {
    id: number;
    name: string;
    serviceType?: string;
    integrationType?: string;
    cutoffTime?: string;
    trackingUrlTemplate?: string;
}

export interface Shipment {
    id?: number;
    outboundOrderId?: number;
    carrierId?: number;
    trackingNumber?: string;
    shippedAt?: string;
    status?: string;
    totalWeightKg?: number;
}

export interface TransportLoad {
    id?: number;
    loadNumber?: string;
    carrierId?: number;
    trailerNumber?: string;
    driverName?: string;
    driverPhone?: string;
    sealNumber?: string;
    status?: string;
    scheduledDeparture?: string;
    actualDeparture?: string;
}

// ===== CARRIERS =====

export const fetchCarriers = async (): Promise<Carrier[]> => {
    const response = await axios.get(`${API_URL}/carriers`);
    return response.data;
};

export const fetchCarrierById = async (id: number): Promise<Carrier> => {
    const response = await axios.get(`${API_URL}/carriers/${id}`);
    return response.data;
};

// ===== SHIPMENTS =====

export const fetchShipments = async (): Promise<Shipment[]> => {
    const response = await axios.get(`${API_URL}/shipments`);
    return response.data;
};

export const fetchShipmentById = async (id: number): Promise<Shipment> => {
    const response = await axios.get(`${API_URL}/shipments/${id}`);
    return response.data;
};

export const createShipment = async (outboundOrderId: number, carrierId: number, trackingNumber?: string): Promise<Shipment> => {
    const response = await axios.post(`${API_URL}/shipments`, null, {
        params: { outboundOrderId, carrierId, trackingNumber }
    });
    return response.data;
};

export const updateShipment = async (id: number, shipment: Partial<Shipment>): Promise<Shipment> => {
    const response = await axios.put(`${API_URL}/shipments/${id}`, shipment);
    return response.data;
};

// ===== LOADS =====

export const fetchLoads = async (): Promise<TransportLoad[]> => {
    const response = await axios.get(`${API_URL}/loads`);
    return response.data;
};

export const createLoad = async (
    carrierId: number,
    trailerNumber?: string,
    driverName?: string,
    driverPhone?: string
): Promise<TransportLoad> => {
    const response = await axios.post(`${API_URL}/loads`, null, {
        params: { carrierId, trailerNumber, driverName, driverPhone }
    });
    return response.data;
};

export const assignShipmentToLoad = async (loadId: number, shipmentId: number): Promise<void> => {
    await axios.post(`${API_URL}/loads/${loadId}/assign/${shipmentId}`);
};

export const dispatchLoad = async (loadId: number): Promise<any> => {
    const response = await axios.post(`${API_URL}/loads/${loadId}/dispatch`);
    return response.data;
};
